homeassistant/components/renault/renault_entities.py

```diff
@@ -0,0 +1,131 @@
+    """Implementation of a Renault entity with a data coordinator."""
+
+    def __init__(
+        self, vehicle: RenaultVehicleProxy, entity_type: str, coordinator_key: str
+    ) -> None:
+        """Initialise entity."""
+        super().__init__(vehicle.coordinators[coordinator_key])
+        self.vehicle = vehicle
+        self._entity_type = entity_type
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return a device description for device registry."""
+        return self.vehicle.device_info
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique identifier for this entity."""
+        return slugify(f"{self.vehicle.details.vin}-{self._entity_type}")
+
+    @property
+    def name(self) -> str:
+        """Return the name of this entity."""
+        return f"{self.vehicle.details.vin}-{self._entity_type}"
+
+    @property
+    def available(self) -> bool:
+        """Return if entity is available."""
+        # Data can succeed, but be empty
+        return self.coordinator.last_update_success and self.coordinator.data
+
+
+class RenaultBatteryDataEntity(RenaultDataEntity):
+    """Implementation of a Renault entity with battery coordinator."""
+
+    def __init__(self, vehicle: RenaultVehicleProxy, entity_type: str) -> None:
+        """Initialise entity."""
+        super().__init__(vehicle, entity_type, "battery")
+
+    @property
+    def data(self) -> KamereonVehicleBatteryStatusData | None:
+        """Return collected data."""
+        return (
+            cast(KamereonVehicleBatteryStatusData, self.coordinator.data)
+            if self.coordinator.data is not None
+            else None
+        )
+
+    @property
+    def device_state_attributes(self) -> dict[str, Any]:
+        """Return the state attributes of this entity."""
+        attrs = {}
+        if self.data is not None and self.data.timestamp is not None:
+            attrs[ATTR_LAST_UPDATE] = self.data.timestamp
+        return attrs
+
+    @property
+    def is_charging(self) -> bool:
+        """Return charge state as boolean."""
+        if self.data is None:
+            return False
+        return self.data.get_charging_status() == ChargeState.CHARGE_IN_PROGRESS
+
+    @property
+    def is_plugged_in(self) -> bool:
+        """Return plug state as boolean."""
+        if self.data is None:
+            return False
+        return self.data.get_plug_status() == PlugState.PLUGGED
+
+
+class RenaultChargeModeDataEntity(RenaultDataEntity):
+    """Implementation of a Renault entity with charge_mode coordinator."""
+
+    def __init__(self, vehicle: RenaultVehicleProxy, entity_type: str) -> None:
+        """Initialise entity."""
+        super().__init__(vehicle, entity_type, "charge_mode")
+
+    @property
+    def data(self) -> KamereonVehicleChargeModeData:
+        """Return collected data."""
+        return cast(KamereonVehicleChargeModeData, self.coordinator.data)
```

Comment:
The return type can be passed into the coordinator, remove the need to cast (TypeVar)

Reply:
Implemented